### PR TITLE
[Bithumb] new exchange implementation

### PR DESCRIPTION
**Pull request by aaron-jang**
Issue #1871

https://www.bithumb.com
API : https://www.bithumb.com/u1/US127

### CODE REVIEW

**Comment by walec51** [comment]
+1 good work

could you add error handling as described in: https://github.com/knowm/XChange/wiki/New-Implementation-Best-Practices#step-7

you can take a look at xchange-bittrex for an example how its done

**Comment by aaron-jang** [comment]
@walec51 

Hi, I added error handling.

**Comment by walec51** [comment]
those try / catches should be in the methods that implement generic interfaces
methods in *Raw classes should throw the exchanges original exception

File: xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/dto/BithumbResponse.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+public class BithumbResponse<D> {
+
+  private final String status;
+  private final String message;
+  private final D data;
+
+  public BithumbResponse(
+      @JsonProperty("status") String status,
+      @JsonProperty("message") String message,
+      @JsonProperty("data") D data) {
```
Comment (walec51) [inline xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/dto/BithumbResponse.java]: if you'd add this here:

```
if (!StringUtils.equals(status, "0000")) {
    throw new ExceptionalReturnContentException("Got error status");
}
```

then  rescu will catch it and re-parse the body as BithumbException

Comment (aaron-jang) [reply xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/dto/BithumbResponse.java]: @walec51 Thanks, I fixed this, too.

File: xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/service/BithumbTradeService.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    try {
+      return BithumbAdapters.adaptOrders(getBithumbOrders(currencyPair));
+    } catch (BithumbException e) {
+      throw BithumbErrorAdapter.adapt(e);
+    }
+  }
+
+  @Override
+  public String placeMarketOrder(MarketOrder marketOrder) throws IOException {
+    return placeBithumbMarketOrder(marketOrder).getOrderId();
```
Comment (walec51) [inline xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/service/BithumbTradeService.java]: try / catch BithumbException needed here

File: xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/service/BithumbTradeService.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  }
+
+  @Override
+  public String placeLimitOrder(LimitOrder limitOrder) throws IOException {
+    return placeBithumbLimitOrder(limitOrder).getOrderId();
```
Comment (walec51) [inline xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/service/BithumbTradeService.java]: try / catch BithumbException needed here

File: xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/service/BithumbTradeService.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  @Override
+  public boolean cancelOrder(CancelOrderParams orderParams) throws IOException {
+
+    if (orderParams instanceof CancelOrderByPairAndIdParams) {
+      final CancelOrderByPairAndIdParams params = (CancelOrderByPairAndIdParams) orderParams;
+      return cancelBithumbOrder(Long.valueOf(params.getOrderId()), params.getCurrencyPair());
+    } else {
+      throw new NotYetImplementedForExchangeException(
+          "Only CancelOrderByPairAndIdParams supported");
+    }
```
Comment (walec51) [inline xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/service/BithumbTradeService.java]: try / catch BithumbException needed here

**Comment by walec51** [comment]
@timmolter looks good to you? can we merge?

**Pull request merged**